### PR TITLE
[LoRA] quality of life improvements in the loading semantics and docs

### PR DESCRIPTION
**Pull request by sayakpaul**
Potentially closes: https://github.com/huggingface/diffusers/issues/3174.

This PR introduces the pieces that were missing in the LoRA docs (general documentation and example READMEs). Sorry for the oversight on my part regarding this. 

I have also amended the loading semantics to cover the different scenarios in which LoRA parameters can be loaded. A rigorous review is appreciated. 

Here are some examples:

**Legacy format**

```py
from diffusers import DiffusionPipeline

model_base = "runwayml/stable-diffusion-v1-5"

pipeline = DiffusionPipeline.from_pretrained(model_base)

lora_model_path = "patrickvonplaten/lora_dreambooth_dog_example"
pipeline.unet.load_attn_procs(lora_model_path)
```

Works but now throws:

image

Also works (note that now we are using the general `load_lora_weights()` method:

```py
from diffusers import DiffusionPipeline

model_base = "runwayml/stable-diffusion-v1-5"

pipeline = DiffusionPipeline.from_pretrained(model_base)

lora_model_path = "patrickvonplaten/lora_dreambooth_dog_example"
pipeline.load_lora_weights(lora_model_path)
```

But throws:

image

**New format**

Fails

```py
from diffusers import DiffusionPipeline

model_base = "runwayml/stable-diffusion-v1-5"

pipeline = DiffusionPipeline.from_pretrained(model_base)

lora_model_path = "sayakpaul/dreambooth"
pipeline.unet.load_attn_procs(lora_model_path)
```

But throws a meaningful error message:

```bash
ValueError: You are using the new LoRA serialization format introduced in 
https://github.com/huggingface/diffusers/pull/2918. Please use `pipe.load_lora_weights(...)`
```

Now, this, of course, works:

```py
from diffusers import DiffusionPipeline

model_base = "runwayml/stable-diffusion-v1-5"

pipeline = DiffusionPipeline.from_pretrained(model_base)

lora_model_path = "sayakpaul/dreambooth"
pipeline.load_lora_weights(lora_model_path)
```

### CODE REVIEW

File: docs/source/en/training/lora.mdx
Comment (sayakpaul) [inline docs/source/en/training/lora.mdx:116]: To better distinguish between the base model path and the LoRA model path.

File: src/diffusers/loaders.py
Comment (sayakpaul) [inline src/diffusers/loaders.py:893]: Made it private to avoid any user confusion.

**Comment by sayakpaul** [comment]
The long [deprecation message](https://github.com/huggingface/diffusers/blob/3045fb276352681f6b9075956e599dd8ef571872/src/diffusers/loaders.py#L835) is not rendered in its complete form:

image

@patrickvonplaten, do you have any suggestions for me to fix it?

File: src/diffusers/loaders.py
Comment (sayakpaul) [inline src/diffusers/loaders.py:893]: Better function naming + private to avoid user confusion.

File: examples/dreambooth/train_dreambooth_lora.py
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:1048]: We should be using the more general method here.

File: src/diffusers/loaders.py
Comment (sayakpaul) [inline src/diffusers/loaders.py:239]: Arises when someone serializes with the new LoRA format and then tries to load with `pipe.unet.load_attn_procs()`.

File: src/diffusers/loaders.py
Comment (sayakpaul) [inline src/diffusers/loaders.py:815]: I had messed up with the filtering 😟 now, it should be fixed.

File: src/diffusers/loaders.py
Comment (sayakpaul) [inline src/diffusers/loaders.py:846]: In case someone doesn't train the text encoder but still, the parameters are serialized in the new format. So, this check is needed.

File: src/diffusers/loaders.py
Comment (patrickvonplaten) [inline src/diffusers/loaders.py:237]: I don't think it's a good idea that this fails. Imagine the following case:

A user does not have a pipeline but just a unet and wants to load the unet lora weights from a "text encoder + unet" serialization format. In this case the natural thing for me would be to extract the unet's lora, throw a warning that a "text encoder + unet" weight is loaded and that the text encoder is disregarded

Comment (patrickvonplaten) [inline src/diffusers/loaders.py:237]: I don't think it's a good idea that this fails. Imagine the following case:

A user does not have a pipeline but just a unet and wants to load the unet lora weights from a "text encoder + unet" serialization format. In this case the natural thing for me would be to extract the unet's lora, throw a warning that a "text encoder + unet" weight is loaded and that the text encoder is disregarded

**Review by patrickvonplaten — COMMENTED** [review]
LoRA is a unique case and we should be careful here about how we design it. We have two function:

`pipe.unet.load_attn_procs(...)` and `pipe.load_lora_weights(...)`. 
In my opinion we need to design the loading in a way that makes sure that we don't block a use case and making sure that we teach the user how to use it correctly. 

We have two serialization formats, let's call them "with `unet.` prefix" (WUP) and "without `unet.` prefix" (WOUP).

Previously I think we said the we want to deprecate the WOUP format, but upon a second thought I actually don't think we can and should deprecated it at all. This is because every time someone has only access to a `unet`, but not a `pipe` they can only save it in the WOUP format  via `save_attn_procs` and it's also expected that `save_attn_procs` saves exactly in the (WOUP) format.

Then, we have two loading functions `pipe.load_lora_weights` and `pipe.unet.load_attn_procs`.

Cool, then we have 4 scenarios (which IMO are **all** important to handle):

1.) WUP + `pipe.load_lora_weights` => Easy, that's expected
2.) WUP + `pipe.unet.load_attn_procs` => This case should IMO work, we should **not** throw an error here. To a big part this is because people might not even have access to the `load_lora_weights` function because they only have the `unet`. Then they want to load a LoRA weight into their unet, but are forced to create a pipeline, which they might not want. So in this case, let's just remove all `text_encoder.` weights, strip the `unet.` and load it and throw a warning that none of the `text_encoder.` weights are being used.
3.) WOUP + `pipe.load_lora_weights` => This is the only case where I'd be willing to throw an error because the user necessarily has access to a `unet` (since we have access to a pipeline). So here we have two options:
a) We throw an error and tell the user to use `f"{self.unet_name}.load_attn_procs(...)` instead.
b) We load the WOUP format, but throw a warning that the user should use `pipe.unet.load_attn_procs` instead because the user can easily access it with `self.unet_name`.

I prefer b) as I think this case is too common to always throw an error and it's also something we allow in `transformers` (e.g. we allow loading `BertModel` weights into the `BertForMaskedLM` class)
4) WOUP + `pipe.unet.load_attn_procs` => I think we shouldn't throw anything here (neither deprecation, nor warning) and just allow it. This is expected behavior and as said above we cannot really stop users from doing it (also we might not want to).

I'd be in favor of rewriting the logic as stated above, so essentially allowing everything and not deprecating anything (a bit different to what we discussed before). Wdyt?

**Comment by sayakpaul** [comment]
I agree with your reasoning 💯. Let me change a few things and get back. 

With the current changes it honestly won't be too hard for me to do that. I mostly reckon working carefully through the depreciation and the warnings.

**Comment by tcapelle** [comment]
I was using the dreambooth_lora script, and this solves the loading:
```python
pipe.load_attn_procs(lora_model_path)
```
but I am unable to use the `cross_attention_kwargs={"scale": 0.5},` anymore, it says that `scale` does not exists. Can you give me info on how to read this?

**Comment by patrickvonplaten** [comment]
> I was using the dreambooth_lora script, and this solves the loading:
> 
> ```python
> pipe.load_attn_procs(lora_model_path)
> ```
> 
> but I am unable to use the `cross_attention_kwargs={"scale": 0.5},` anymore, it says that `scale` does not exists. Can you give me info on how to read this?

Hey @tcapelle, 

Could you maybe open a new issue for this as it's not really related? :-) Thanks a lot!

**Comment by tcapelle** [comment]
Ok, just to confirm about the loading:
- When Training a LoRA with the [text_to_image_lora](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_lora.py) script, I can load the model back by using: `pipe.unet.load_attn_procs(lora_model_path)`
- When trainina LoRA+dreambooth using [train_dreambooth_lora](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora.py) script I can load the model back using: `pipe.load_attn_procs(lora_model_path)` but I can't use the `scale` param anymore. Also, the model that is produced is bad. It actually ignores the LoRa.
Would you mind [making this project](https://wandb.ai/patrickvonplaten/dreambooth-lora/reports/LoRA-DreamBooth-Example--VmlldzozMzY0MTgz) public @patrickvonplaten ?

**Comment by sayakpaul** [comment]
A> When trainina LoRA+dreambooth using train_dreambooth_lora script I can load the model back using: pipe.load_attn_procs(lora_model_path) but I can't use the scale param anymore. Also, the model that is produced is bad. It actually ignores the LoRa. 

I don't think you are using the latest changes introduced in this PR. I would suggest either you revert to the commit that was previously working for you or wait for this PR to get merged as it will address all of your concerns except the `scale` parameter. Once this PR is merged, I welcome you to try it out again and report the findings. 

I think having the issues this way will be easier for us to track :)

**Comment by tcapelle** [comment]
Thanks Paul =)

File: examples/dreambooth/train_dreambooth_lora.py
Comment (patrickvonplaten) [reply examples/dreambooth/train_dreambooth_lora.py:1048]: Agree!

**Review by patrickvonplaten — APPROVED** [review]
Awesome!

File: src/diffusers/loaders.py
Comment (patrickvonplaten) [inline src/diffusers/loaders.py:93]: Think this is the best option we have right now! 

Just FYI: Feel slightly uncomfortable with the unet having to know about the text encoder name here as we now super lightly entangle those concepts now, but I also think that this is the best design at the moment! Thank for adding it!

**Review by patrickvonplaten — APPROVED** [review]

File: src/diffusers/loaders.py
Comment (sayakpaul) [reply src/diffusers/loaders.py:93]: Do you want me to add the names to the utils constants?

File: docs/source/en/training/lora.mdx
Comment (sayakpaul) [inline docs/source/en/training/lora.mdx:247]: Since we decided not to deprecate things, I changed the wording here. @patrickvonplaten as an FYI.

**Pull request merged**
